File: app/src/main/java/com/websarva/wings/android/zuboradiary/data/database/DiaryItemTitleSelectionHistoryRepository.java
package com.websarva.wings.android.zuboradiary.data.database;

import androidx.annotation.NonNull;

import com.google.common.util.concurrent.ListenableFuture;

import java.util.List;
import java.util.Objects;

import javax.inject.Inject;

public class DiaryItemTitleSelectionHistoryRepository {
    private final DiaryItemTitleSelectionHistoryDAO diaryItemTitleSelectionHistoryDAO;

    @Inject
    public DiaryItemTitleSelectionHistoryRepository(DiaryItemTitleSelectionHistoryDAO diaryItemTitleSelectionHistoryDAO) {
        Objects.requireNonNull(diaryItemTitleSelectionHistoryDAO);

        this.diaryItemTitleSelectionHistoryDAO = diaryItemTitleSelectionHistoryDAO;
    }

    @NonNull
    public ListenableFuture<List<DiaryItemTitleSelectionHistoryItemEntity>> loadSelectionHistory(
            int num, int offset) {
        if (num < 1) throw new IllegalArgumentException();
        if (offset < 0) throw new IllegalArgumentException();

        ListenableFuture<List<DiaryItemTitleSelectionHistoryItemEntity>> future =
                diaryItemTitleSelectionHistoryDAO.selectHistoryListOrderByLogDesc(num, offset);
        return Objects.requireNonNull(future);
    }

    // MEMO:保存する時は日記保存と同時に処理したいので、DiaryRepositoryにて処理。
    @NonNull
    public ListenableFuture<List<Long>> saveSelectionHistoryItems(
                                            List<DiaryItemTitleSelectionHistoryItemEntity> list) {
        Objects.requireNonNull(list);
        list.stream().forEach(Objects::requireNonNull);

        ListenableFuture<List<Long>> future = diaryItemTitleSelectionHistoryDAO.insertHistoryItem(list);
        return Objects.requireNonNull(future);
    }

    @NonNull
    public ListenableFuture<Integer> deleteSelectionHistoryItem(String title) {
        Objects.requireNonNull(title);

        ListenableFuture<Integer> future = diaryItemTitleSelectionHistoryDAO.deleteHistoryItem(title);
        return Objects.requireNonNull(future);
    }

    // MEMO:保存する時は日記保存と同時に処理したいので、DiaryRepositoryにて処理。
    @NonNull
    public ListenableFuture<Integer> deleteOldHistoryItems() {
        ListenableFuture<Integer> future = diaryItemTitleSelectionHistoryDAO.deleteOldHistoryItem();
        return Objects.requireNonNull(future);
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/data/database/DiaryItemTitleSelectionHistoryRepository.kt
package com.websarva.wings.android.zuboradiary.data.database

import com.websarva.wings.android.zuboradiary.data.mapper.toDomainModel
import com.websarva.wings.android.zuboradiary.domain.model.DiaryItemTitleSelectionHistoryItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext

internal class DiaryItemTitleSelectionHistoryRepository (
    private val diaryItemTitleSelectionHistoryDAO: DiaryItemTitleSelectionHistoryDAO
) {

    fun loadSelectionHistory(
        num: Int, offset: Int
    ): Flow<List<DiaryItemTitleSelectionHistoryItem>> {
        require(num >= 1)
        require(offset >= 0)

        return diaryItemTitleSelectionHistoryDAO
            .selectHistoryListOrderByLogDesc(num, offset)
            .map { list ->
                list.map { it.toDomainModel() }
            }
    }

    suspend fun deleteSelectionHistoryItem(title: String) {
        withContext(Dispatchers.IO) {
            diaryItemTitleSelectionHistoryDAO.deleteHistoryItem(title)
        }
    }
}
